components/encryption/src/log_file.rs

```diff
@@ -0,0 +1,394 @@
+                        }
+                        LogRecord::REMOVE => {
+                            let original = file_dict.files.remove(&file_name);
+                            if original.is_none() {
+                                return Err(box_err!(
+                                    "Try to recovery from log file but remove a null entry, file name: {}",
+                                    file_name
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(file_dict)
+    }
+
+    /// Append an insert operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn insert(&mut self, name: &str, info: FileInfo) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes = Self::convert_record_to_bytes(name, LogRecord::INSERT(info))?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
+            self.check_compact()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+
+    /// Append a remove operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn remove(&mut self, name: &str) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes = Self::convert_record_to_bytes(name, LogRecord::REMOVE)?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
+            self.check_compact()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+
+    /// Replace existed file entry with new file information.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn replace(&mut self, src_name: &str, dst_name: &str, info: FileInfo) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes1 = Self::convert_record_to_bytes(src_name, LogRecord::REMOVE)?;
+            let bytes2 = Self::convert_record_to_bytes(dst_name, LogRecord::INSERT(info))?;
+            file.write_all(&bytes1)?;
+            file.write_all(&bytes2)?;
+            file.sync_all()?;
+            self.check_compact()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+
+    /// This function needs to be called after each append operation to check
+    /// if compact is needed.
+    fn check_compact(&mut self) -> Result<()> {
+        self.append_num += 1;
+        if self.append_num > Self::REWRITE_THRESHOLD {
+            let file_dict = self.recovery()?;
```

Comment:
Recovery it is expensive. So, why not keep a internal `FileDictionary`? So the recovery can be omit.
And, `self.write` will block all updates about `LogFile`. It's not friendly. You can change the disk file layout to
```
ACTIVE_DICT_FILE
OLD_FILE_3
OLD_FILE_2
OLD_FILE_1
```
Then write code like this:
```rust
if too_much_garbage_in(self.active_dict_file) {
    rename(self.active_dict_file, gen_next_inactive_file_name());
    let old_file_dict = self.old_file_dict.take();
    std::spawn(move || rewrite_old_files(old_file_dict, get_all_old_file_names());
}
```

Reply:
refer to https://github.com/tikv/tikv/pull/8865#issuecomment-716969238, reverted the background thread commit.